File: crates/sovereign_wallet/src/protocol.rs
//! Framed protocol types for `openagents.omega.sovereign-wallet.v1` (design §2).

use serde::{Deserialize, Serialize};
use serde_json::Value;

pub const PROTOCOL_SCHEMA: &str = "openagents.omega.sovereign-wallet.v1";
pub const PROTOCOL_VERSION: u32 = 1;
pub const SERVICE_VERSION: &str = "0.1.0";
/// Newline-framed JSON frames must stay under this byte budget (design §2.1).
pub const MAX_FRAME_BYTES: usize = 64 * 1024;

/// wavecli-style stable error codes (design §2.3).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ErrorCode {
    InvalidArgs,
    WalletNotCreated,
    WalletLocked,
    WalletSyncing,
    NotFound,
    MethodNotFound,
    ConfirmationRequired,
    InsufficientBalance,
    InvoiceExpired,
    Canceled,
    DeadlineExceeded,
    Aborted,
    WaitTimeout,
    MainnetRefused,
    PaymentHashMismatch,
    CredentialConsumed,
    StaleGeneration,
    AlreadyRunning,
    WavedBinaryMissing,
    WavedWalletApiUnavailable,
    IncompatibleVersion,
    Internal,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ErrorEnvelope {
    pub code: ErrorCode,
    pub message: String,
    #[serde(default)]
    pub details: String,
    #[serde(default)]
    pub retryable: bool,
    #[serde(default)]
    pub remediation: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RequestFrame {
    pub schema: String,
    pub kind: String,
    pub id: String,
    pub generation: u64,
    pub method: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub params: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResponseFrame {
    pub schema: String,
    pub kind: String,
    pub id: String,
    pub generation: u64,
    pub ok: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub result: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<ErrorEnvelope>,
}

/// Wallet state normalized from the Wavelength daemon (design §1.4).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum WalletState {
    None,
    Locked,
    Syncing,
    Ready,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InitializeResult {
    pub schema: String,
    pub protocol_version: u32,
    pub service_version: String,
    pub generation: u64,
    pub capabilities: Vec<String>,
    pub data_root: String,
    pub network: String,
    /// The waved daemon's ACTUAL runtime network from the probe (SEC-2026-050),
    /// never the env value alone. `None` when waved is unavailable.
    #[serde(default)]
    pub waved_network: Option<String>,
    #[serde(default)]
    pub waved_state: String,
    #[serde(default)]
    pub http_surface: HttpSurface,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct HttpSurface {
    #[serde(default)]
    pub bound: bool,
    #[serde(default)]
    pub port: u16,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HealthResult {
    pub ok: bool,
    pub status: String,
    pub generation: u64,
    pub data_root: String,
    pub wallet_state: WalletState,
    pub waved_connected: bool,
    pub network: String,
    #[serde(default)]
    pub note: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StatusResult {
    pub schema: String,
    pub protocol_version: u32,
    pub service_version: String,
    pub network: String,
    pub waved_connected: bool,
    #[serde(default)]
    pub waved_unavailable_reason: String,
    pub wallet_state: WalletState,
    #[serde(default)]
    pub vault_state: String,
    #[serde(default)]
    pub l402_gateway_state: String,
    pub http_surface: HttpSurface,
    pub data_root: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BalanceResult {
    pub confirmed_sat: String,
    pub pending_in_sat: String,
    pub pending_out_sat: String,
    pub credit_available_sat: String,
    pub credit_reserved_sat: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InvoiceResult {
    pub invoice: String,
    #[serde(default)]
    pub payment_hash: Option<String>,
    pub amount_sat: u64,
    #[serde(default)]
    pub memo: String,
    #[serde(default)]
    pub hrp: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PayResult {
    #[serde(default)]
    pub payment_hash: Option<String>,
    #[serde(default)]
    pub status: String,
    #[serde(default)]
    pub activity_id: String,
    #[serde(default)]
    pub actual_amount_sat: String,
    #[serde(default)]
    pub expected_fee_sat: String,
    #[serde(default)]
    pub fee_known: bool,
    #[serde(default)]
    pub warning: String,
    /// Proof of payment; the sidecar redacts it from every log and never
    /// persists it (SEC-2026-046/047).
    #[serde(default)]
    pub preimage: Option<String>,
}

pub fn request_frame(id: impl Into<String>, generation: u64, method: &str, params: Option<Value>) -> RequestFrame {
    RequestFrame {
        schema: PROTOCOL_SCHEMA.to_string(),
        kind: "request".to_string(),
        id: id.into(),
        generation,
        method: method.to_string(),
        params,
    }
}

File: sidecar/sovereign-wallet/src/protocol.ts
//! Framed loopback protocol for `openagents.omega.sovereign-wallet.v1`.
//!
//! Adopted from the verified omega_effectd frame discipline (one JSON object
//! per newline, bounded frames, generation fencing) with the wavecli-style
//! error envelope (`code`/`message`/`details`/`retryable`/`remediation`).
//! Design: 02-wp1-system-design.md §2.

export const PROTOCOL_SCHEMA = "openagents.omega.sovereign-wallet.v1";
export const PROTOCOL_VERSION = 1;
export const SERVICE_VERSION = "0.1.0";
/** Newline-framed JSON frames must stay under this byte budget (design §2.1). */
export const MAX_FRAME_BYTES = 64 * 1024;

/** Stable machine-readable error codes (wavecli-style; design §2.3). */
export type ErrorCode =
  | "INVALID_ARGS"
  | "WALLET_NOT_CREATED"
  | "WALLET_LOCKED"
  | "WALLET_SYNCING"
  | "NOT_FOUND"
  | "METHOD_NOT_FOUND"
  | "CONFIRMATION_REQUIRED"
  | "INSUFFICIENT_BALANCE"
  | "INVOICE_EXPIRED"
  | "CANCELED"
  | "DEADLINE_EXCEEDED"
  | "ABORTED"
  | "WAIT_TIMEOUT"
  | "MAINNET_REFUSED"
  | "PAYMENT_HASH_MISMATCH"
  | "CREDENTIAL_CONSUMED"
  | "STALE_GENERATION"
  | "ALREADY_RUNNING"
  | "EXPORT_ALREADY_CONSUMED"
  | "WAVED_BINARY_MISSING"
  | "WAVED_WALLET_API_UNAVAILABLE"
  | "INCOMPATIBLE_VERSION"
  | "INTERNAL";

export interface ErrorEnvelope {
  code: ErrorCode;
  message: string;
  details: string;
  retryable: boolean;
  remediation: string;
}

export interface RequestFrame {
  schema: string;
  kind: "request";
  id: string;
  generation: number;
  method: string;
  params?: unknown;
}

export interface ResponseFrame {
  schema: string;
  kind: "response";
  id: string;
  generation: number;
  ok: boolean;
  result?: unknown;
  error?: ErrorEnvelope;
}

export interface EventFrame {
  schema: string;
  kind: "event";
  id: "0";
  generation: number;
  method: string;
  params: unknown;
}

export type Frame = RequestFrame | ResponseFrame | EventFrame;

export function errorEnvelope(
  code: ErrorCode,
  message: string,
  opts: { details?: string; retryable?: boolean; remediation?: string } = {},
): ErrorEnvelope {
  return {
    code,
    message,
    details: opts.details ?? "",
    retryable: opts.retryable ?? false,
    remediation: opts.remediation ?? "",
  };
}

/** Methods that move funds must never be blindly retried after a timeout (wavecli rule). */
export function isBlindRetryUnsafe(code: ErrorCode): boolean {
  return (
    code === "DEADLINE_EXCEEDED" ||
    code === "ABORTED" ||
    code === "WAIT_TIMEOUT" ||
    code === "CANCELED"
  );
}

export function encodeResponse(frame: ResponseFrame): string {
  return `${JSON.stringify(frame)}\n`;
}

export function encodeEvent(generation: number, method: string, params: unknown): string {
  const frame: EventFrame = {
    schema: PROTOCOL_SCHEMA,
    kind: "event",
    id: "0",
    generation,
    method,
    params,
  };
  return `${JSON.stringify(frame)}\n`;
}